libp2p/connection.nim

```diff
@@ -66,51 +65,57 @@ proc `$`*(conn: Connection): string =
   if not isNil(conn.peerInfo):
     result = $(conn.peerInfo)
 
-proc bindStreamClose(conn: Connection) {.async.} =
-  # bind stream's close event to connection's close
-  # to ensure correct close propagation
-  if not isNil(conn.stream.closeEvent):
-    await conn.stream.closeEvent.wait()
-    trace "wrapped stream closed, about to close conn",
-      closed = conn.isClosed, conn = $conn
-    if not conn.isClosed:
-      trace "wrapped stream closed, closing conn",
-        closed = conn.isClosed, conn = $conn
-      await conn.close()
-
 proc init[T: Connection](self: var T, stream: LPStream): T =
   ## create a new Connection for the specified async reader/writer
   new self
   self.stream = stream
-  self.closeEvent = newAsyncEvent()
-  when chronicles.enabledLogLevel == LogLevel.TRACE:
-    self.oid = genOid()
-  asyncCheck self.bindStreamClose()
-  inc getConnectionTracker().opened
-  libp2p_open_connection.inc()
-
+  self.initStream()
   return self
 
 proc newConnection*(stream: LPStream): Connection =
   ## create a new Connection for the specified async reader/writer
   result.init(stream)
 
+method initStream*(s: Connection) =
+  procCall LPStream(s).initStream()
+  trace "created connection", oid = s.oid
+  inc getConnectionTracker().opened
+  libp2p_open_connection.inc()
+
 method readExactly*(s: Connection,
                     pbytes: pointer,
                     nbytes: int):
-                    Future[void] {.gcsafe.} =
-  s.stream.readExactly(pbytes, nbytes)
+                    Future[void] {.async, gcsafe.} =
+  try:
+    await s.stream.readExactly(pbytes, nbytes)
+  except CatchableError as exc:
+    await s.close()
```

Comment:
:scream: wasn't it supposed to be the caller that was the owner of the connection and therefore closed it on error?

Reply:
Yeah this puzzled me too, it's against our current design I thought

Reply:
so this needs to be written down somewhere, what the expectations are, around raising, ownership and closing - ie documented in `init` or `close` or somewhere, so that one can make the right adjustments when fixing bugs in the future - for example, if every level catches, does cleanup, and raises, that has implications for how the cleanup should be propagated across the object hierarchy - who cleans up who and when - so that it's done in a well-defined order and doesn't race.

Reply:
This is a hard one, both from a conceptual perspective as well as implementation wise. Ideally, I would have a working `raises` pragma that allows me to patch up all the places where `EOF` or `Closed` is thrown, and dispose of the connection there, but I don't so this is the next best thing. BTW, we can still make this work without this ugly workaround, but there is a good chance that sometime in the future, where another read is introduced somewhere else, we'll again start to leak connections. Open to suggestions as always.